### PR TITLE
Fees API Improvement

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

## Problem

The current API provided by the traits around `pallet_fees` to handle fees has two negative points:
1. In order to use it in a pallet you need to use at least 2 associated types in the `Config` trait. i.e:
  ```rust
  /// Entity used to pay fees
  type Fees: Fees<AccountId = Self::AccountId>;

  /// Key used to get the fee value.
  type ActionAFeeKey: Get<<Self::Fees as Fees>::FeeKey>;
  ```
2. The knowledge of how to consume that fee is in the pallet when the pallet should not be aware of it. This should be a runtime concern.
  ```rust
  /// Inside the pallet we decide when we should not do that
  T::Fees::fee_to_treasury(&who, Fee::Key(T::ActionAFeeKey))?;
  // or
  T::Fees::fee_to_author(&who, Fee::Balance(T::ActionAFeeValue))?;
  ```
  
## Solution

We can fix these 2 points by coding the whole fee action in different types (see PR code) under a new trait `PayFee`.

Pallet:
```rust
// In Config trait:
type ActionAPayFee: PayFee<T::Account>;

// When use it:
T::ActionAPayFee::pay_for(&who)?;
```

Runtime:
```rust
parameter_types! {
    pub const ActionAFee: Fee = Fee::Balance(10 * CFG);
}

impl pallet_example::Config for Runtime {
    // We set in one line:
    //   - How the fee is handled
    //   - The value of the fee
    type ActionAPayFee = FeeToTreasury<Fees, ActionAFee>;
    // ...
}
```

### CODE REVIEW

**Comment by lemunozm** [comment]
This PR is completed and ready for review

**Review by cdamian — APPROVED** [review]

**Review by NunoAlexandre — DISMISSED** [review]
Sweet 🍯 👌

File: libs/traits/src/lib.rs [hunk 1]
```diff
@@ -435,6 +435,13 @@ pub mod fees {
 			fee: Fee<Self::Balance, Self::FeeKey>,
 		) -> DispatchResult;
 	}
+
+	/// Trait to pay fees
+	/// This trait can be used by pallet to just pay fees without worring about
+	/// the value or where the fee goes.
+	pub trait PayFee<AccountId> {
+		fn pay_for(who: &AccountId) -> DispatchResult;
```
[the remark this answers is not in the document]
Comment (lemunozm) [reply libs/traits/src/lib.rs:443]: Ok, makes sense!

[the remark this answers is not in the document]
Comment (lemunozm) [reply libs/traits/src/lib.rs]: Nah, never mind, I've updated the PR with `pay()`

File: libs/traits/src/lib.rs [hunk 2]
```diff
@@ -435,6 +435,13 @@ pub mod fees {
 			fee: Fee<Self::Balance, Self::FeeKey>,
 		) -> DispatchResult;
 	}
+
+	/// Trait to pay fees
+	/// This trait can be used by pallet to just pay fees without worring about
+	/// the value or where the fee goes.
+	pub trait PayFee<AccountId> {
+		fn pay(who: &AccountId) -> DispatchResult;
```
Comment (NunoAlexandre) [inline libs/traits/src/lib.rs:443]: Given the previous discussion, I would for sure document that `who` is the account that's paying for the fees and not the account that is receiving them.

**Review by NunoAlexandre — APPROVED** [review]
One doc request but not blocking 🙏

File: libs/traits/src/lib.rs (hunk 2 above)
Comment (wischli) [reply libs/traits/src/lib.rs:443]: +1 on that. Would also be great to rename `who` to `payer`.

Comment (lemunozm) [reply libs/traits/src/lib.rs:443]: Good point! Sorry for the "auto-merge", keep in mind for a future PR where I could add this 🧠

Comment (lemunozm) [reply libs/traits/src/lib.rs:443]: Here it goes: https://github.com/centrifuge/centrifuge-chain/pull/1629/commits/2901f0ec4d758715b5bbdc89feba843590496944

**Pull request merged**